### PR TITLE
Support Stream module's server_name directive by using SNI extension o…

### PR DESCRIPTION
**Pull request by mrpre**
# Primary Changes
* Support `Stream` module's `server_name` directive 
It's semantic equivalence between `HTTP` module and `Stream` module except that `Stream` module only uses SNI extension of TLS to locate server block.  
It gives Tengine an ability to distribute the `Stream` request by another dimension rather than using different port.  

* `server_name` only works under SSL offload which means  `server_name` only works with `listen ... ssl;`.  
*  Support `Stream` module's `ssl_sni_force` directive(Default is off) 
Reject the TLS handshake if SNI doesn't match the `server_name` we configured.

### CODE REVIEW

File: src/stream/ngx_stream.c [hunk 1]
```diff
@@ -523,6 +740,15 @@ ngx_stream_optimize_servers(ngx_conf_t *cf, ngx_array_t *ports)
 
             stport->naddrs = i + 1;
 
+#if (NGX_STREAM_SNI)
+            /*Because of ssl_sni_force we have to add hash key even one server*/
+            if (addr[i].servers.nelts >= 1) {
+                if (ngx_stream_server_names(cf, addr->opt.ctx->main_conf[ngx_stream_core_module.ctx_index], &addr[i]) != NGX_OK) {
```
[the remark this answers is not in the document]
Comment (mrpre) [reply src/stream/ngx_stream.c]: fixed

File: src/stream/ngx_stream_ssl_module.c [hunk 2]
```diff
@@ -44,6 +44,11 @@ static char *ngx_stream_ssl_session_cache(ngx_conf_t *cf, ngx_command_t *cmd,
     void *conf);
 static ngx_int_t ngx_stream_ssl_init(ngx_conf_t *cf);
 
+#if (NGX_STREAM_SNI)
+int
+ngx_stream_ssl_servername(ngx_ssl_conn_t *ssl_conn, int *ad, void *arg);
```
[the remark this answers is not in the document]
Comment (mrpre) [reply src/stream/ngx_stream_ssl_module.c]: fixed

File: tests/nginx-tests/nginx-tests/stream_sni.t [hunk 3]
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/perl
+
+# (C) Sergey Kandaurov
+# (C) Nginx, Inc.
```
Comment (chobits) [inline tests/nginx-tests/nginx-tests/stream_sni.t]: fix to right Author

File: src/stream/ngx_stream.c [hunk 4]
```diff
@@ -444,7 +661,9 @@ ngx_stream_optimize_servers(ngx_conf_t *cf, ngx_array_t *ports)
     ngx_stream_conf_port_t      *port;
     ngx_stream_conf_addr_t      *addr;
     ngx_stream_core_srv_conf_t  *cscf;
-
+#if (NGX_STREAM_SNI)
+    ngx_stream_core_main_conf_t *cmcf;
+#endif
```
Comment (chobits) [inline src/stream/ngx_stream.c]: coding style: dont remove this one space

File: src/stream/ngx_stream_ssl_module.c [hunk 5]
```diff
@@ -193,6 +198,15 @@ static ngx_command_t  ngx_stream_ssl_commands[] = {
       offsetof(ngx_stream_ssl_conf_t, crl),
       NULL },
 
+#if (NGX_STREAM_SNI)
+    { ngx_string("ssl_sni_force"),
+        NGX_STREAM_MAIN_CONF|NGX_STREAM_SRV_CONF|NGX_CONF_FLAG,
```
Comment (chobits) [inline src/stream/ngx_stream_ssl_module.c]: coding style: align to no.202 line

File: tests/nginx-tests/nginx-tests/stream_sni.t (hunk 3 above)
Comment (mrpre) [reply tests/nginx-tests/nginx-tests/stream_sni.t]: fixed

File: src/stream/ngx_stream.c (hunk 4 above)
Comment (mrpre) [reply src/stream/ngx_stream.c]: fixed

File: src/stream/ngx_stream_ssl_module.c (hunk 5 above)
Comment (mrpre) [reply src/stream/ngx_stream_ssl_module.c]: fixed

**Comment by chobits** [comment]
ACK the logic.

Note that this feature is experimental.  We'll deprecate this feature if there is any conflict with similar feature of nginx official.

**Comment by wangfakang** [comment]
ACK.

File: src/stream/ngx_stream_ssl_module.c [hunk 6]
```diff
@@ -636,6 +653,10 @@ ngx_stream_ssl_merge_conf(ngx_conf_t *cf, void *parent, void *child)
 
     ngx_conf_merge_str_value(conf->ciphers, prev->ciphers, NGX_DEFAULT_CIPHERS);
 
+#if (NGX_STREAM_SNI)
+    ngx_conf_merge_value(conf->sni_force, prev->sni_force, 0);
+    if (!conf->listen)
```
Comment (chobits) [inline src/stream/ngx_stream_ssl_module.c:658]: wrong added line ` if (!conf->listen)`, it make conf->ssl.log NULL.

Then the following nginx.conf will trigger a segfault when nginx starts.

```
stream {
    server {
        listen 127.0.0.1:8980 reuseport ssl udp;
        ssl_protocols dtlsv1;
        ssl_certificate_key localhost.key;
        ssl_certificate localhost.crt;
        return "ok";
    }
}
```

Comment (chobits) [reply src/stream/ngx_stream_ssl_module.c:658]: fixed in https://github.com/alibaba/tengine/pull/1684

**Pull request merged**